### PR TITLE
30ignition: add support for ignition-disks

### PR DESCRIPTION
**Pull request by lucab**
This adds all relevant bits to run Ignition "disks" stage.

### CODE REVIEW

**Comment by lucab** [comment]
I tested this with a secondary blank disk (sda) on libvirt with this CLC YAML: https://gist.github.com/lucab/64eb8e622e4444b20cad296c98554803

Partitioning + mkfs + files writing confirmed working. I didn't check RAID support: creation should work out of the box but assembling logic is distro-specific (in CL, it has its own separated GUID part-type detection and udev logic)

File: dracut/30ignition/ignition-files.service [hunk 1]
```diff
@@ -6,11 +6,12 @@ Before=initrd-parse-etc.service
 Requires=initrd-root-fs.target
 After=initrd-root-fs.target
 
-#Requires=ignition-setup.service ignition-disks.service
-#After=ignition-setup.service ignition-disks.service
+#Requires=ignition-setup.service
+#After=ignition-setup.service
```
Comment (dustymabe) [inline dracut/30ignition/ignition-files.service]: should we just delete the above two lines?

Comment (lucab) [reply dracut/30ignition/ignition-files.service]: I'd say not for now, there are still many cleanups that we can do in a single sweep once we have stabilized the whole initramfs sequence. But if I need to respin this PR for another round I can drop them there, yes.

**Comment by dustymabe** [comment]
i'm seeing an error in fedora (i changed your config to use sdb instead because that is my 2nd disk)

```
    9.263461] ignition[767]: Ignition 0.28.0
[    9.264455] systemd[1]: Stopped target Slices.
[    9.265496] ignition[767]: reading system config file "/usr/lib/ignition/base.ign"[  OK  ] Started Cleaning Up and Shutting Down Daemons.

[    9.267480] systemd[1]: Stopped target Sockets.
[    9.268545] ignition[767]: no config at "/usr/lib/ignition/base.ign"
[    9.269785] systemd[1]: Stopped target Paths.
[    9.271019] systemd[1]: Stopped dracut mount hook.
[    9.272335] ignition[767]: no config URL provided
[    9.273502] systemd[1]: Stopped dracut pre-mount hook.
[    9.274754] ignition[767]: reading system config file "/usr/lib/ignition/user.ign"
[    9.276289] systemd[1]: Stopped dracut initqueue hook.
[    9.277707] ignition[767]: no config at "/usr/lib/ignition/user.ign"
[    9.279220] systemd[1]: Stopping Open-iSCSI...
[    9.280406] ignition[767]: op(1): [started]  loading QEMU firmware config module
[    9.281987] audit: type=1131 audit(1537387336.132:22): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=dracut-mount comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'                                                                                                           
[    9.285588] audit: type=1130 audit(1537387336.137:23): pid=1 uid=0 auid=4294967295 ses=4294967295 subj=kernel msg='unit=dracut-pre-mount comm="systemd" exe="/usr/lib/systemd/systemd" hostname=? addr=? terminal=? res=success'
[    9.289418] systemd[1]: Started Cleaning Up and Shutting Down Daemons.
[    9.290867] ignition[767]: op(1): [finished] loading QEMU firmware config module[FAILED] Failed to start Ignition (disks).

See 'systemctl status ignition-disks.service' for details.
[    9.294936] [DEPEND] Dependency failed for Ignition (files).
[DEPEND] Dependency failed for Switch Root.
[DEPEND] Dependency failed for Switch Root.
systemd[1]: ignition-disks.service: Main process exited, code=exited, status=1/FAILURE
         Starting Setup Virtual Console...
         Starting Ask OSTree to mount /var in initramfs...
[    9.302734] ignition[767]: disks failed                                                                                                                                                                                                                                                                                                
[    9.307807] systemd[1]: ignition-disks.service: Failed with result 'exit-code'.
[    9.308969] ignition[767]: disks: createPartitions: op(2): [started]  waiting for devices [/dev/sdb]
[    9.310341] systemd[1]: Failed to start Ignition (disks).
[    9.311309] ignition[767]: disks: createPartitions: op(2): [failed]   waiting for devices [/dev/sdb]: dbus: authentication protocol error
[[    9.313098] systemd[1]: Dependency failed for Ignition (files).
[    9.313458] ignition[767]: Ignition failed: create partitions failed: failed to wait on disks devs: dbus: authentication protocol error
[    9.313801] systemd[1]: Dependency failed for Switch Root.
[    9.314528] iscsid[498]: iscsid shutting down.
[    9.314875] systemd[1]: Dependency failed for Switch Root.
```

I used ignition rpms from here to test with: http://koji.fedoraproject.org/koji/taskinfo?taskID=29765014

I'll test with RHCOS now and see if I see the same thing there.

**Comment by lucab** [comment]
@dustymabe vanilla systemd v239? Then it's likely https://github.com/systemd/systemd/issues/9553. For reference, we are still at v238 on CL due to multiple regressions.

**Comment by dustymabe** [comment]
> @dustymabe vanilla systemd v239? Then it's likely [systemd/systemd#9553](https://github.com/systemd/systemd/issues/9553). For reference, we are still at v238 on CL due to multiple regressions.

yeah it's the `239` in fedora 29. don't know if `vanilla` is the right word to use but probably does not include the fix: https://github.com/systemd/systemd/pull/9625.. should we try to get that backported to fedora? 

the good news is that if you don't try to use ignition-disks then it doesn't break anything. 

I'll try to build RHCOS and see what I get.

**Comment by dustymabe** [comment]
ok finally got RHCOS built.. I'm still seeing an issue where ignition-disks fails because /dev/sdb1 never shows up:

<details>

```
[    8.360475] ignition[1811]: disks: createPartitions: op(2): [started]  waiting for devices [/dev/sdb]                                                                                                                                    
[    8.362548] systemd[1]: Closed Open-iSCSI iscsiuio Socket.                                                                                                                                                                               
[    8.364670] iscsid[439]: iscsid shutting down.                                                                                                                                                                                           
[    8.366077] systemd[1]: Stopped target System Initialization.                                                                                                                                                                            
[    8.367735] ignition[1811]: disks: createPartitions: op(2): [finished] waiting for devices [/dev/sdb]                                                                                                                                    
[    8.369931] systemd[1]: Stopped target Local File Systems.                                                                                                                                                                               
[    8.371499] ignition[1811]: disks: createPartitions: created device alias for "/dev/sdb": "/dev_aliases/dev/sdb" -> "/dev/sdb"                                                                                                           
[    8.374050] systemd[1]: Stopped target Swap.                                                                                                                                                                                             
[    8.375370] ignition[1811]: disks: createPartitions: op(3): [started]  partitioning "/dev_aliases/dev/sdb"                                                                                                                               
[    8.377644] systemd[1]: Stopped target Local Encrypted Volumes.                                                                                                                                                                          
[    8.379272] ignition[1811]: disks: createPartitions: op(3): wiping partition table requested on "/dev_aliases/dev/sdb"                                                                                                                   
[    8.381734] systemd[1]: Stopped Apply Kernel Variables.                                                                                                                                                                                  
[    8.383217] ignition[1811]: disks: createPartitions: op(3): running sgdisk with options: [--zap-all /dev_aliases/dev/sdb]                                                                                                                
[    8.385715] systemd[1]: Stopped udev Coldplug all Devices.                                                                                                                                                                               
[    8.387250] ignition[1811]: disks: createPartitions: op(3): op(4): [started]  deleting 0 partitions and creating 0 partitions on "/dev_aliases/dev/sdb"                                                                                  
[    8.390146] systemd[1]: Stopped dracut pre-trigger hook.                                           
[    8.392157] systemd[1]: Stopping udev Kernel Device Manager...                                                                                                                                                                          
[    8.393759] systemd[1]: Stopped target Slices.                   
[    8.395431] systemd[1]: Stopped udev Kernel Device Manager.           
[    8.396989] systemd[1]: Stopped Open-iSCSI.                                
[    8.398178] systemd[1]: Stopped Create Static Device Nodes in /dev.    
[    8.399668] systemd[1]: Stopped Create list of required static device nodes for the current kernel.                                                                                                                                     
[    8.401646] systemd[1]: Stopped dracut pre-udev hook.
[    8.402929] systemd[1]: Stopped dracut cmdline hook.
[    8.404176] systemd[1]: Closed udev Kernel Socket.                                                                                                                                                                                      
[    8.405409] systemd[1]: Closed udev Control Socket.                                                                                                                                                                                     
[    8.406688] systemd[1]: Starting Cleanup udevd DB...
[    8.407954] systemd[1]: Started Cleanup udevd DB.                      
[    8.409180] systemd[1]: Reached target Switch Root.
[    9.437770] ignition[1811]: disks: createPartitions: op(3): op(4): [finished] deleting 0 partitions and creating 0 partitions on "/dev_aliases/dev/sdb"
[    9.445106] ignition[1811]: disks: createPartitions: op(3): op(5): [started]  reading partition table of "/dev_aliases/dev/sdb"
[    9.447965] ignition[1811]: disks: createPartitions: op(3): op(5): [finished] reading partition table of "/dev_aliases/dev/sdb"
[    9.450477] ignition[1811]: disks: createPartitions: op(3): running sgdisk with options: [--pretend --new=0:0:+0 /dev_aliases/dev/sdb]
[    9.452794] ignition[1811]: disks: createPartitions: op(3): running sgdisk with options: [--new=0:0:+0 --change-name=0:EXTRAGPT0 /dev_aliases/dev/sdb]                                                                                  
[    9.455446] ignition[1811]: disks: createPartitions: op(3): op(6): [started]  deleting 0 partitions and creating 1 partitions on "/dev_aliases/dev/sdb"
[   10.483098]  sdb: sdb1                                      
[   10.484940] ignition[1811]: disks: createPartitions: op(3): op(6): [finished] deleting 0 partitions and creating 1 partitions on "/dev_aliases/dev/sdb"
[   10.490290] ignition[1811]: disks: createPartitions: op(3): [finished] partitioning "/dev_aliases/dev/sdb"
[   10.493119] ignition[1811]: disks: createFilesystems: op(7): [started]  waiting for devices [/dev/sdb1]                                                                                                                                  
[   68.155087] random: crng init done                                                                                                                                                                                                      
[  100.511411] systemd[1]: Job dev-sdb1.device/start timed out.
[  100.514447] systemd[1]: Timed out waiting for device dev-sdb1.device.                                                                                                                                                                    
[  100.517943] ignition[1811]: disks: createFilesystems: op(7): [failed]   waiting for devices [/dev/sdb1]: device unit dev-sdb1.device timeout                                                                                             
[  100.524446] systemd[1]: Job dev-sdb1.device/start failed with result 'timeout'.                                                                                                                                                         
[  100.526651] ignition[1811]: Ignition failed: failed to create filesystems: failed to wait on filesystems devs: device unit dev-sdb1.device timeout
[  100.529099] systemd[1]: ignition-disks.service: main process exited, code=exited, status=1/FAILURE
[  100.531344] ignition[1811]: disks failed
[  100.533433] systemd[1]: Failed to start Ignition (disks).
```
</details>

when I hit the emergency shell I do see `/dev/sdb1`. 

```
:/# blkid /dev/sdb1
/dev/sdb1: PARTLABEL="EXTRAGPT0" PARTUUID="97255054-4ea6-4130-8f10-68d7d3c3acbe"
```

**Comment by lucab** [comment]
Meh :(

Partitioning did indeed work and you can see the kernel picks up the new table (`sdb: sdb1`). However this info doesn't reach userland in time. Can you please check if it eventually makes it (`udevadm info /dev/sdb1 | grep TAGS`)? Also which platform and kind of disk is that?

Options are:
 * the default timeout is too short -> we can try to tweak https://www.freedesktop.org/software/systemd/man/systemd.unit.html#JobTimeoutSec=
 * udev needs an additional poke -> an `udevadm settle` between partitioning and mkfs may help
 * older systemd/udev worked in a different way -> this depends on RHEL details that I don't know

/cc @ajeddeloh as the second item likely makes sense to do on anyway

**Comment by dustymabe** [comment]
> Partitioning did indeed work and you can see the kernel picks up the new table (`sdb: sdb1`). However this info doesn't reach userland in time. Can you please check if it eventually makes it (`udevadm info /dev/sdb1 | grep TAGS`)? Also which platform and kind of disk is that?

running on libvirt/qemu on Fedora 27. - here is the snippet from libvirt xml for the secondary disk (sdb):

```
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/guests/storagepools/manual/foo.img16396.2'/>
      <target dev='hdb' bus='ide'/>
      <address type='drive' controller='0' bus='0' target='0' unit='1'/>
    </disk>
```


`udevadm info`:

```
:/# udevadm info /dev/sdb1 | grep TAGS
:/# udevadm info /dev/sdb1 
P: /devices/pci0000:00/0000:00:01.1/ata1/host0/target0:0:1/0:0:1:0/block/sdb/sdb1
N: sdb1
E: DEVNAME=/dev/sdb1
E: DEVPATH=/devices/pci0000:00/0000:00:01.1/ata1/host0/target0:0:1/0:0:1:0/block/sdb/sdb1
E: DEVTYPE=partition
E: MAJOR=8
E: MINOR=17
E: SUBSYSTEM=block
```

**Comment by lucab** [comment]
It looks like some udev mismatch, as all the other symlinks are missing too and there is no systemd tag. Comparing to my test, I think the main difference is that I'm using a `virtio` backend while you are using plain `ide`.

---

Disk stanza:
<details>

```
    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/lib/libvirt/images/vm-rhcos-pre.qcow2'/>
      <backingStore/>
      <target dev='sda' bus='scsi'/>
      <alias name='scsi0-0-0-0'/>
      <address type='drive' controller='0' bus='0' target='0' unit='0'/>
    </disk>
```

</details>

---

Udev entry from initramfs, after a reboot+`rd.break`:
<details>

```
# udevadm info /dev/sda1
P: /devices/pci0000:00/0000:00:09.0/virtio5/host2/target2:0:0/2:0:0:0/block/sda/sda1
N: sda1
S: disk/by-id/scsi-0QEMU_QEMU_HARDDISK_drive-scsi0-0-0-0-part1
S: disk/by-label/EXTRAFS0
S: disk/by-partlabel/EXTRAGPT0
S: disk/by-partuuid/59de0361-f70f-4a5d-904d-451834c271b7
S: disk/by-path/pci-0000:00:09.0-scsi-0:0:0:0-part1
S: disk/by-uuid/07c6f42a-a04f-4680-bc90-91c48d9c6a6c
E: DEVLINKS=/dev/disk/by-id/scsi-0QEMU_QEMU_HARDDISK_drive-scsi0-0-0-0-part1 /dev/disk/by-label/EXTRAFS0 /dev/disk/by-partlabel/EXTRAGPT0 /dev/disk/by-partuuid/59de0361-f70f-4a5d-904d-451834c271b7 /dev/disk/by-path/pci-0000:00:09.0-scsi-0:0:0:0-part1 /dev/disk/by-uuid/07c6f42a-a04f-4680-bc90-91c48d9c6a6c
E: DEVNAME=/dev/sda1
E: DEVPATH=/devices/pci0000:00/0000:00:09.0/virtio5/host2/target2:0:0/2:0:0:0/block/sda/sda1
E: DEVTYPE=partition
E: ID_BUS=scsi
E: ID_FS_LABEL=EXTRAFS0
E: ID_FS_LABEL_ENC=EXTRAFS0
E: ID_FS_TYPE=ext4
E: ID_FS_USAGE=filesystem
E: ID_FS_UUID=07c6f42a-a04f-4680-bc90-91c48d9c6a6c
E: ID_FS_UUID_ENC=07c6f42a-a04f-4680-bc90-91c48d9c6a6c
E: ID_FS_VERSION=1.0
E: ID_MODEL=QEMU_HARDDISK
E: ID_MODEL_ENC=QEMU\x20HARDDISK\x20\x20\x20
E: ID_PART_ENTRY_DISK=8:0
E: ID_PART_ENTRY_NAME=EXTRAGPT0
E: ID_PART_ENTRY_NUMBER=1
E: ID_PART_ENTRY_OFFSET=2048
E: ID_PART_ENTRY_SCHEME=gpt
E: ID_PART_ENTRY_SIZE=2095071
E: ID_PART_ENTRY_TYPE=0fc63daf-8483-4772-8e79-3d69d8477de4
E: ID_PART_ENTRY_UUID=59de0361-f70f-4a5d-904d-451834c271b7
E: ID_PART_TABLE_TYPE=gpt
E: ID_PATH=pci-0000:00:09.0-scsi-0:0:0:0
E: ID_PATH_TAG=pci-0000_00_09_0-scsi-0_0_0_0
E: ID_REVISION=2.5+
E: ID_SCSI=1
E: ID_SERIAL=0QEMU_QEMU_HARDDISK_drive-scsi0-0-0-0
E: ID_SERIAL_SHORT=drive-scsi0-0-0-0
E: ID_TYPE=disk
E: ID_VENDOR=QEMU
E: ID_VENDOR_ENC=QEMU\x20\x20\x20\x20
E: MAJOR=8
E: MINOR=1
E: SUBSYSTEM=block
E: TAGS=:systemd:
E: USEC_INITIALIZED=46476
```

</details>

---

Relevant rules:

<details>

```
udevadm test /devices/pci0000:00/0000:00:09.0/virtio5/host2/targt2:0:0/2:0:0:0/block/sda/sda1
[...]
GROUP 0 /usr/lib/udev/rules.d/50-udev-default.rules:52
LINK 'disk/by-id/scsi-0QEMU_QEMU_HARDDISK_drive-scsi0-0-0-0-part1' /usr/lib/udev/rules.d/60-persistent-storage.rules:54
LINK 'disk/by-path/pci-0000:00:09.0-scsi-0:0:0:0-part1' /usr/lib/udev/rules.d/60-persistent-storage.rules:69
IMPORT builtin 'blkid' /usr/lib/udev/rules.d/60-persistent-storage.rules:87
probe /dev/sda1 raid offset=0
LINK 'disk/by-uuid/07c6f42a-a04f-4680-bc90-91c48d9c6a6c' /usr/lib/udev/rules.d/60-persistent-storage.rules:93
LINK 'disk/by-label/EXTRAFS0' /usr/lib/udev/rules.d/60-persistent-storage.rules:94
LINK 'disk/by-partuuid/59de0361-f70f-4a5d-904d-451834c271b7' /usr/lib/udev/rules.d/60-persistent-storage.rules:101
LINK 'disk/by-partlabel/EXTRAGPT0' /usr/lib/udev/rules.d/60-persistent-storage.rules:102
```

</details>

**Comment by ajeddeloh** [comment]
Couple things:
* In terms of raid, the special rules are only for doing root-on-raid, which we're not doing in this case, so it should "just work". Be aware it does no magic though, adding units to start raid is up to the user.
* `/dev/sdb1` never showing up is almost certainly a system/kernel issue. Adding a settle probably wont fix that since that basically just blocks until udev has nothing to do. If anything a poke with `sgdisk -p` might be more effective.
* I'm guessing if it's not showing up in the default timeout it's not showing up at all.

**Comment by dustymabe** [comment]
ok i switched to scsi and IT WORKED the first time, but then reverted back and failed subsequent boots in the same was as for IDE. we've definitely got a race condition or something going on here.

here is the log from the one time it "worked":


```
[   10.494692] ignition[1820]: disks: createFilesystems: op(9): [started]  creating "ext4" filesystem on "/dev_aliases/dev/sdb1"
[   10.577633] ignition[1820]: disks: createFilesystems: op(9): [finished] creating "ext4" filesystem on "/dev_aliases/dev/sdb1"
[   10.582998] ignition[1820]: disks: op(a): [started]  waiting for udev to settle
[   10.585686] systemd[1]: Started Ignition (disks).
[   10.587089] ignition[1820]: disks: op(a): [finished] waiting for udev to settle
[   10.588920] ignition[1820]: disks: disks passed
[   10.590081] systemd[1]: Starting Ask OSTree to mount /var in initramfs...
[   10.592175] ignition[1820]: Ignition finished successfully
[   10.593822] systemd[1]: Started Ask OSTree to mount /var in initramfs.
[   10.595835] systemd[1]: Starting OSTree Prepare OS/...
[   10.600019] ostree-prepare-root[1839]: Resolved OSTree target to: /sysroot/ostree/deploy/rhcos/deploy/4a1d7a1b748aafddf9d5787e525c4da3f12f8ecfd13b6f43f25c1666562ccaa8.0                                                                                                   
[   10.603738] systemd[1]: Started OSTree Prepare OS/.
[   10.605412] systemd[1]: Starting Ignition (files)...
[   10.610954] ignition[1841]: INFO     : Ignition 0.28.0
[   10.612741] ignition[1841]: INFO     : reading system config file "/usr/lib/ignition/base.ign"
[   10.615151] ignition[1841]: INFO     : no config at "/usr/lib/ignition/base.ign"
[   10.623577] ignition[1841]: INFO     : files: createUsers: op(1): [started]  creating or modifying user "core"
[   10.625827] ignition[1841]: DEBUG    : files: createUsers: op(1): executing: "/usr/sbin/usermod" "--root" "/sysroot" "core"
[   10.629859] ignition[1841]: INFO     : files: createUsers: op(1): [finished] creating or modifying user "core"
[   10.632105] ignition[1841]: INFO     : files: createUsers: op(2): [started]  adding ssh keys to user "core"
[   10.634261] ignition[1841]: INFO     : files: createUsers: op(2): [finished] adding ssh keys to user "core"
[   10.636424] ignition[1841]: INFO     : files: createFilesystemsFiles: createFiles: op(3): [started]  mounting "/dev/sdb1" at "/tmp/ignition-files113746923"                                                                                                                
[   10.667476] EXT4-fs (sdb1): mounted filesystem with ordered data mode. Opts: (null)
[   10.669962] ignition[1841]: INFO     : files: createFilesystemsFiles: createFiles: op(3): [finished] mounting "/dev/sdb1" at "/tmp/ignition-files113746923"                                                                                                                
[   10.672406] ignition[1841]: INFO     : files: createFilesystemsFiles: createFiles: op(4): [started]  writing file "/dummy-file"
[   10.675243] ignition[1841]: INFO     : files: createFilesystemsFiles: createFiles: op(4): [finished] writing file "/dummy-file"
[   10.677799] ignition[1841]: INFO     : files: createFilesystemsFiles: createFiles: op(5): [started]  unmounting "/dev/sdb1" at "/tmp/ignition-files113746923"                                                                                                              
[   10.690395] ignition[1841]: INFO     : files: createFilesystemsFiles: createFiles: op(5): [finished] unmounting "/dev/sdb1" at "/tmp/ignition-files113746923"                                                                                                              
[   10.695146] ignition[1841]: INFO     : files: op(6): [started]  processing unit "ignition-relabel.service"
[   10.698386] systemd[1]: Started Ignition (files).
[   10.700844] ignition[1841]: INFO     : files: op(6): op(7): [started]  writing unit "ignition-relabel.service" at "run/systemd/system/ignition-relabel.service"                                                                                                            
[   10.703648] random: crng init done
[   10.704393] ignition[1841]: INFO     : files: op(6): op(7): [finished] writing unit "ignition-relabel.service" at "run/systemd/system/ignition-relabel.service"                                                                                                            
[   10.707077] ignition[1841]: INFO     : files: op(6): [finished] processing unit "ignition-relabel.service"
[   10.711475] ignition[1841]: INFO     : files: files passed
[   10.712711] ignition[1841]: INFO     : Ignition finished successfully
[   10.714134] systemd[1]: Starting Switch Root...
[   10.715399] systemd[1]: Switching root.
[   10.716466] systemd-journald[119]: Received SIGTERM from PID 1 (n/a).
[   10.735674] type=1404 audit(1537467952.968:2): enforcing=1 old_enforcing=0 auid=4294967295 ses=4294967295
[   10.804235] SELinux:  Class bpf not defined in policy.
[   10.805157] SELinux: the above unknown classes and permissions will be allowed
[   10.814513] type=1403 audit(1537467953.047:3): policy loaded auid=4294967295 ses=4294967295
[   10.820332] systemd[1]: Successfully loaded SELinux policy in 84.336ms.
[   10.834681] ip_tables: (C) 2000-2006 Netfilter Core Team
[   10.835688] systemd[1]: Inserted module 'ip_tables'
[   10.851767] systemd[1]: Relabelled /dev, /run and /sys/fs/cgroup in 13.240ms.
[   10.854549] systemd[1]: [/etc/systemd/system.conf.d/origin-accounting.conf:5] Unknown lvalue 'DefaultIOAccounting' in section 'Manager'                                                                                                                                    
[   11.311839] systemd-journald[1879]: Received request to flush runtime journal from PID 1
[   11.666556] piix4_smbus 0000:00:01.3: SMBus Host Controller at 0x700, revision 0
[   11.702730] sd 0:0:0:0: Attached scsi generic sg0 type 0
[   11.703858] sd 0:0:1:0: Attached scsi generic sg1 type 0
[   11.705124] sd 2:0:0:0: Attached scsi generic sg2 type 0
[   11.706936] sd 2:0:0:1: Attached scsi generic sg3 type 0
[   11.808853] input: PC Speaker as /devices/platform/pcspkr/input/input4
[   11.854406] Error: Driver 'pcspkr' is already registered, aborting...
[   12.069142] ppdev: user-space parallel port driver

Cannot open access to console, the root account is locked.
See sulogin(8) man page for more details.

Press Enter to continue.
```

**Comment by lucab** [comment]
@dustymabe from the log, it looks like even on failure the disks stage and the whole initramfs was successful. It sounds like agetty is detecting something broken on the system, so I think it is not the same problem you were seeing with IDE. Are you using my config or do you have some additional stuff in there (e.g. a mount unit)?

@ajeddeloh the device is showing up (kernel-side), but udev/systemd are not detecting it. If it happens sometime for any kind of disk, it could be a race. If it happens reliably only for IDE, it could be an issue with udev rules.

**Comment by lucab** [comment]
I tried switching to IDE disks and triggering a few cycles of disks provisioning (both with blank and reused disks) and I still haven't encountered this.
As this could be anything from cosmic rays, to a specific race to some oddity on how we are locally stitching together an image, I'd suggest to start landing the initial service unit in RHCOS and recheck if there any oddities there.

**Comment by dustymabe** [comment]
> @dustymabe from the log, it looks like even on failure the disks stage and the whole initramfs was successful.

that was the log from the one time it worked.. all of the failure logs look like the log from https://github.com/dustymabe/ignition-dracut/pull/17#issuecomment-423007853


> As this could be anything from cosmic rays, to a specific race to some oddity on how we are locally stitching together an image, I'd suggest to start landing the initial service unit in RHCOS and recheck if there any oddities there.

yeah I don't mind landing this since people won't hit it unless they tell ignition to do something with it. the problem I see is that both places I tried to get it to work haven't worked. Fedora 29 issues with systemd 239 and EL7 with some sort of race condition. I linked you to the exact qcow i'm using in chat, can you see if you can reproduce with that qcow?

**Comment by lucab** [comment]
@dustymabe your image does indeed fail at mkfs "most of the times".

It seems too reliable to be a race, and it doesn't depend on disk type, but some other factor made it working a couple of time for me (with no apparent environment variations). Kernel is happy with picking up the new partition table, but udev sees it only after a `systemctl restart systemd-udev-trigger.service` (all other kinds of poking previously suggested failed at it though).

https://github.com/openshift/os/pull/294 happened in parallel with this PR, and I fear some units dependencies/ordering got dropped or re-arranged because of that. I'll start from scratch again from latest RHCOS image and re-check the service graph. Please hold this PR in the meanwhile.

**Comment by dustymabe** [comment]
> Please hold this PR in the meanwhile.

+1

**Comment by lucab** [comment]
So, I re-tried on latest RHCOS (`4.0.5930`) and did a few changes:
 * re-ordered disks stage to be sandwiched between `basic` and `initrd-root-fs` targets (see [diagram](https://www.freedesktop.org/software/systemd/man/bootup.html)) in order to have proper udevd/network/etc
 * removed direct dependency between ignition stages, as they happen on two different sides of `initrd-cleanup`

Tested with the same CLC as before, with multiple combinations of IDE/SCSI, blank and re-used. I didn't see any failure locally.

Additionally, a backport request for Fedora systemd v239 is up at https://bugzilla.redhat.com/show_bug.cgi?id=1631840.

File: dracut/30ignition/ignition-files.service [hunk 2]
```diff
@@ -21,13 +18,13 @@ After=ostree-prepare-root.service
 After=initrd-switch-root.target
 Before=initrd-switch-root.service
 
+## NOTE(lucab): this was the original order in CL. This stage now runs
+##  after initrd-cleanup, and some behavior may be subtly different.
+
+#Before=initrd-parse-etc.service
```
Comment (dustymabe) [inline dracut/30ignition/ignition-files.service]: should this comment and line be moved back up to line 4 in the file where initrd-parse-etc.service was before?

File: dracut/30ignition/ignition-disks.service [hunk 3]
```diff
@@ -2,13 +2,17 @@
...
-Requires=sysinit.target
-After=sysinit.target
+# This stage runs between `basic.target` and `initrd-root-fs.target`,
+# see https://www.freedesktop.org/software/systemd/man/bootup.html
+After=basic.target
+Before=initrd-root-fs.target
 
-Requires=network.target
+# Network may be used to fetch userdata content.
 After=network.target
```
Comment (dustymabe) [inline dracut/30ignition/ignition-disks.service]: no need for `Requires=network.target` ?

File: dracut/30ignition/ignition-files.service (hunk 2 above)
Comment (lucab) [reply dracut/30ignition/ignition-files.service]: All of those can disappear once we are done with https://github.com/dustymabe/ignition-dracut/issues/20.

File: dracut/30ignition/ignition-disks.service (hunk 3 above)
Comment (lucab) [reply dracut/30ignition/ignition-disks.service]: That's a tricky topic. We don't strictly require networking (eg. libvirt with no additional remote URLs), but if networking is going to be available we prefer to be started after that. I will revert back to the original unit, as it seems less surprise prone.

Comment (dustymabe) [reply dracut/30ignition/ignition-disks.service]: yeah I think `network.target` is a bit confusing anyway. 

from https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/:

- *network.target has very little meaning during start-up. It only indicates that the network management stack is up after it has been reached.*
- *network-online.target is a target that actively waits until the nework is "up"*

**Comment by dustymabe** [comment]
LGTM

**Pull request merged**